core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -717,41 +721,40 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
 		};
 
 		Box::new(move |who, intent, topic, mut data| {
-			if let MessageIntent::PeriodicRebroadcast = intent {
-				return do_rebroadcast;
-			}
-
-			let peer = match inner.peers.peer(who) {
-				None => return false,
-				Some(x) => x,
-			};
-
-			// if the topic is not something we're keeping at the moment,
-			// do not send.
-			let (maybe_round, set_id) = match inner.live_topics.topic_info(&topic) {
-				None => return false,
-				Some(x) => x,
-			};
-
-			// if the topic is not something the peer accepts, discard.
-			if let Some(round) = maybe_round {
-				return peer.view.consider_vote(round, set_id) == Consider::Accept
-			}
-
-			// global message.
-			let our_best_commit = inner.local_view.last_commit;
-			let peer_best_commit = peer.view.last_commit;
-
+			
 			match GossipMessage::<Block>::decode(&mut data) {
```

Comment:
IMO we want to avoid re-decoding (the fact that we have to do it at all is an unfortunate side-effect of the API). Any checks that can be done without decoding the message should be done before attempting to decode. AFAIK, round and neighbor messages can be determined before decoding.